core/src/main/java/com/linecorp/armeria/internal/common/GracefulConnectionShutdownHandler.java

```diff
@@ -0,0 +1,98 @@
+package com.linecorp.armeria.internal.common;
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+/**
+ * Abstract class that's used implement protocol-specific graceful connection shutdown logic.
+ */
+public abstract class GracefulConnectionShutdownHandler {
+    private static final Logger logger = LoggerFactory.getLogger(GracefulConnectionShutdownHandler.class);
+
+    @Nullable
+    ChannelPromise promise;
+    Duration gracePeriod = Duration.ZERO;
+    boolean started;
+    @Nullable
+    private ScheduledFuture<?> gracePeriodFuture;
+
+    /**
+     * Code executed on grace period start. Executed at most once.
+     */
+    public abstract void onGracePeriodStart(ChannelHandlerContext ctx);
+    /**
+     * Code executed on grace period end. Executed at most once.
+     */
+    public abstract void onGracePeriodEnd(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception;
+
+    public void start(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        if (this.promise == null) {
+            this.promise = promise;
+        } else {
+            // Chain promises in case start is called multiple times.
+            this.promise.addListener((ChannelFutureListener) future -> {
+                if (future.cause() != null) {
+                    promise.setFailure(future.cause());
+                } else {
+                    promise.setSuccess();
+                }
+            });
+        }
+        if (gracePeriodFuture != null) {
+            if (gracePeriodFuture.getDelay(TimeUnit.NANOSECONDS) > gracePeriod.toNanos()) {
+                // Maybe reschedule below.
+                gracePeriodFuture.cancel(false);
+                gracePeriodFuture = null;
+            } else {
+                // Grace period is already scheduled to finish earlier.
+                return;
+            }
+        }
+        if (gracePeriod.compareTo(Duration.ZERO) > 0) {
```

Comment:
Maybe cheaper to use `gracePeriod.toNanos() > 0` or how about just storing everything in millis rather than in `Duration`?

Reply:
I'm not sure about this

- Duration promotes a more readable API 
- Users may need micros precision (e.g. round trip within the same DC is ~500 microseconds)
- By quickly looking at the sources - toNanos seems to involve multiplication, while compare simply compares 2 numbers, it should be faster. But I don't think we should worry about the performance in this case, it's not going to be on the critical path and not going to be executed frequently, readability and ability to have higher precision than millis is more important IMO.

Reply:
They are good point indeed, but here's where I'm coming from: https://github.com/line/armeria/pull/3715#discussion_r677044859

Reply:
Discussed offline, updated to store drain period as long internally to make sure it doesn't cause GC issues in cases where Armeria needs to close large number of connections.